GH-43809: [Docs] Update extension type examples to not use UUID

khwilson opened the pull request:
### Rationale for this change

UUID extension types were made canonical in #41299 and are getting native support in C++ and Python in #37298. As such, it makes sense to provide an alternative user-defined extension type as an example that is unlikely to become a canonical extension type anytime soon.

After discussion in #43809, we determined a `RationalType` would make sense.

### What changes are included in this PR?

A change in several doc locations which reference a `UuidType` extension type have been changed to a `RationalType`.

For consistency, this PR also changes single quotes (`''`) to double quotes (`""`) throughout the Python examples that it modifies.

### Are these changes tested?

I've run all the code in the docs locally. I'm not sure if Arrow runs doctests, but welcome information on how to enable those for these changes.

### Are there any user-facing changes?

No.

cc @ianmcook @rok 
* GitHub Issue: #43809

ianmcook commented:
@khwilson I found one more docs page where there are a couple of mentions of UUID as an example of a user-defined extension type:

https://github.com/apache/arrow/blob/75ca5b3631144f58ea3edbe6b4933a686c0e0fd9/docs/source/format/Columnar.rst?plain=1#L1600-L1604

https://github.com/apache/arrow/blob/75ca5b3631144f58ea3edbe6b4933a686c0e0fd9/docs/source/format/Columnar.rst?plain=1#L1610

Could you replace these with your rational example? Thanks

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
-            assert storage_type == pa.binary(16)
+            assert pt.is_struct(storage_type)
+            assert pt.is_int32(storage_type[0].type)
             assert serialized == b''
```
```suggestion
            assert serialized == b""
```

khwilson replied:
A question: There are several other places in this file where single quotes were used, e.g., https://github.com/apache/arrow/blob/fde32158fa5f9a5f943da5aacd2bd8056b7a5bb1/docs/source/python/extending_types.rst?plain=1#L267

Would it be helpful to add an issue to standardize to double quotes?

ianmcook replied:
> Would it be helpful to add an issue to standardize to double quotes?

Sure, or if you just want to fix them in this PR, that's fine too. I don't think we're too fussy about consistency of single/double quotes in examples, but it's best to at least be consistent within one example.

khwilson replied:
Fair! I'll go ahead and change them all while I'm here.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
 
         @classmethod
-        def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        def __arrow_ext_deserialize__(self, storage_type, serialized):
```
```suggestion
        def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
             # Sanity checks, not required but illustrate the method signature.
-            assert storage_type == pa.binary(16)
+            assert pt.is_struct(storage_type)
+            assert pt.is_int32(storage_type[0].type)
```
Might as well check denom too
```suggestion
            assert pt.is_int32(storage_type[0].type)
            assert pt.is_int32(storage_type[1].type)
```

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
+be represented as a pair of integers::
 
-    class UuidType(pa.ExtensionType):
+    import pyarrow as pa
+    import pyarrow.types as pt
+
```
I think it's better to:
- omit these imports from the example
- just use `pa.types` in the code (instead of `pt`) here and in the other files

khwilson replied:
Done

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
+
 
 The special methods ``__arrow_ext_serialize__`` and ``__arrow_ext_deserialize__``
-define the serialization of an extension type instance. For non-parametric
-types such as the above, the serialization payload can be left empty.
```
Why remove this?

ianmcook replied:
Oh, I see. I think this relates to what exactly parameterized / parametric means, correct?

- If the storage type requires one or more parameters to initially specify, does that qualify as "parameterized"?
- Or does "parameterized" mean that we need additional metadata (that is not encoded in the underlying storage type) in order to reconstruct the instance?

I think you are defining it to mean the latter, correct?

I think that makes sense. But it would be good to explicitly say this in the docs. See my other comment about that.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
+    StructType(struct<numer: int32, denom: int32>)
+
+    >>> storage_array = pa.array(
+    ... [
+    ...     {"numer": 10, "denom": 17},
+    ...     {"numer": 20, "denom": 13},
+    ... ],
+    ... type=rational_type.storage_type
```
Please indent (like in the example in `types.pxi` below)

khwilson replied:
Done!

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
+    ...             "my_package.rational",
+    ...         )
+    ...     def __arrow_ext_serialize__(self) -> bytes:
+    ...         # No serialized metadata necessary
+    ...         return b""
```
This might be confusing to some readers. You're saying that even though this extension type is "parameterized" (in a sense), it does not need serialized metadata because it's self-describing (i.e. the parameters are known by looking at the storage type), correct? That is a subtle distinction that might confuse readers.

Could you take a stab at adding a sentence or two to describe this distinction in the section under the `Parameterized extension type` heading in `extending_types.rst` https://arrow.apache.org/docs/python/extending_types.html#parameterized-extension-type

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
+    ...     ],
+    ...     type=rational_type.storage_type
+    ... )
+    >>> ratoinal_type.wrap_array(storage_array)
```
Small improvement to prevent readers from mistakenly thinking that `wrap_array` modifies `storage_array` in-place.
```suggestion
    >>> rational_array = ratoinal_type.wrap_array(storage_array)
    >>> rational_array
```

khwilson replied:
Done.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -131,58 +131,82 @@ and serialization mechanism. The extension name and serialized metadata
+    ... ],
+    ... type=rational_type.storage_type
+    ... )
+    >>> arr = rational_type.wrap_array(storage_array)
+    >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
```
```suggestion
    >>> arr = rational_type.wrap_array(storage_array)
    # or equivalently
    >>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
```

khwilson replied:
This is the same as the original documentation. I'm guessing the previous writer didn't insert `# or equivalently` as as it causes `doctest` to fail. I've changed this to the following:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)
>>> # or equivalently
>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

But the alternative (which I think is harder to read) is:

```python
"""
>>> arr = rational_type.wrap_array(storage_array)

or equivalently

>>> arr = pa.ExtensionArray.from_storage(rational_type, storage_array)
"""
```

ianmcook replied:
Thanks. That looks good to me.

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
     Or do the same with creating an ExtensionArray:
 
-    >>> pa.ExtensionArray.from_storage(uuid_type, storage_array)
+    >>> pa.ExtensionArray.from_storage(rational_type, storage_array)
```
```suggestion
    >>> rational_array = pa.ExtensionArray.from_storage(rational_type, storage_array)
    >>> rational_array
```

khwilson replied:
Done

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
+    ...         # No serialized metadata necessary
+    ...         return b""
+    ...     @classmethod
+    ...     def __arrow_ext_deserialize__(self, storage_type, serialized):
```
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

khwilson replied:
fwiw, the `self` is actually in the function signature for `__arrow_ext_deserialize__` on L1753. I've modified that signature as well :-)

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -2089,30 +2120,41 @@ def unregister_extension_type(type_name):
+    ...         # No serialized metadata necessary
+    ...         return b""
+    ...     @classmethod
+    ...     def __arrow_ext_deserialize__(self, storage_type, serialized):
```
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

khwilson replied:
Done

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -2039,30 +2059,41 @@ def register_extension_type(ext_type):
+    ...         # No serialized metadata necessary
+    ...         return b""
+    ...     @classmethod
+    ...     def __arrow_ext_deserialize__(self, storage_type, serialized):
```
```suggestion
    ...     def __arrow_ext_deserialize__(cls, storage_type, serialized):
```

khwilson replied:
Done

ianmcook commented:
@khwilson thank you; this looks pretty good! I left a bunch of comments. They should be mostly be easy to address (except the bit about what "parameterized" means which might require some thinking about how best to explain it).

khwilson commented:
Thanks for the review, @ianmcook! I've addressed all the minor comments and will take another pass at the "parameterized" definition on the cross continental flight I'm about to get on.

khwilson commented:
Apologies for my delay, @ianmcook . Ended up just taking a nap on that flight. :-)

I've taken a stab at disambiguating "parameterization." To do so, I decided to explicitly call the metadata captured by the `ARROW:extension:metadata` field "parameters." If this is outside the bounds of a PR and should be discussed more broadly, please let me know.

The only remaining unaddressed comment from your review about the UUID type in Columnar.rst. However, I actually don't read that one as needing to be replaced _unless_ the intention of _canonical_ extension types is that all Arrow implementations _must_ or _should_ support them. Reading the page on [canonical extensions](https://arrow.apache.org/docs/format/CanonicalExtensions.html) it is unclear to me if arrow implementations _must_ , _should_ , or _may_ support them. If it is _may_, then I think the text of Columnar is actually quite good, though I might modify it as follows:

> For example, Arrow specifies a canonical extension type that represents a UUID as a ``FixedSizeBinary(16)``. Arrow implementations are not required to support canonical extensions, so an implementation that does not support this UUID type will simply interpret it as a ``FixedSizeBinary(16)`` and pass along the ``custom_metadata`` in subsequent Arrow protocol messages.

ianmcook commented:
Thanks @khwilson! I will re-review later this week.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
+to :ref:`Python scalars <custom-scalar-conversion>` and
+:ref:`pandas <conversion-to-pandas>`) and to more easily interpret data.
+
+In the Arrow :doc:`metadata specification</format/Columnar.html#format-metadata-extension-types>`,
```
Please use the existing ref. 
```suggestion
In the Arrow :ref:`format_metadata_extension_types`,
```

This points here:
https://github.com/apache/arrow/blob/ab0a40ee34217070f14027776682074c55d0b507/docs/source/format/Columnar.rst?plain=1#L1575-L1578

khwilson replied:
Thanks! Missed that reference.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
-and serialization mechanism. The extension name and serialized metadata
-can potentially be recognized by other (non-Python) Arrow implementations
+Arrow affords a notion of extension types which allow users to annotate data
+types with additional semantics. This allows downstream consumers both to
```
I'm -1 on the "downstream consumers" phrasing. How about just "developers"?

khwilson replied:
Fair. I had in mind that end users might also come to rely on metadata (e.g., you might use this extension type to create an SPSS-like survey data format), but this seems like a reasonably niche use case.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -197,19 +227,44 @@ and then reading it back yields the proper type::
+    RationalType(StructType(struct<numer: int32, denom: int32>))
+
+Further, note that while we registered the concrete type
+``RationalType(pa.int32())``, ``RationalType(integer_type)`` has the same
+extension name (``"my_package.rational"``) for all integer types. As such,
+the above code will also allow users to (de)serialize these data types::
```
Edit for readability
```suggestion
``RationalType(pa.int32())``, you could use the same extension name
``"my_package.rational"`` to register ``RationalType(integer_type)`` for
any Arrow integer type. As such, the above code will also allow users to
(de)serialize these data types::
```

khwilson replied:
Adopted with a small change.

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -116,73 +116,103 @@ a :class:`~pyarrow.Array` or a :class:`~pyarrow.ChunkedArray`.
-            # Since we don't have a parameterized type, we don't need extra
-            # metadata to be deserialized
-            return b''
+Pyarrow allows you to define extension types from Python by subclassing
```
```suggestion
PyArrow allows you to define extension types from Python by subclassing
```

ianmcook commented on docs/source/python/extending_types.rst:
```diff
@@ -225,22 +280,22 @@ of the given frequency since 1970.
             # attributes need to be set first before calling
             # super init (as that calls serialize)
             self._freq = freq
-            super().__init__(pa.int64(), 'my_package.period')
+            super().__init__(pa.int64(), "my_package.period")
 
         @property
         def freq(self):
             return self._freq
 
         def __arrow_ext_serialize__(self):
-            return "freq={}".format(self.freq).encode()
+            return "freq={self.freq}".encode()
```
Please revert the change to this one line. It was correct before but now it is not.

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,78 @@ cdef class ExtensionType(BaseExtensionType):
+    >>> pa.unregister_extension_type("my_package.rational")
+
+    Note that even though we registered the concrete type
+    ``RationalType(pa.int64())``, pyarrow will be able to deserialize
+    ``RationalType(integer_type)`` for any ``integer_type`` as the deserializer
```
```suggestion
    ``RationalType(pa.int64())``, PyArrow will be able to deserialize
    ``RationalType(integer_type)`` for any ``integer_type``, as the deserializer
```

ianmcook commented:
@khwilson I made some suggested changes. Once you've incorporated these I'll do a final pass and then get some help to figure out why the CI is still failing.

ianmcook commented:
> If this is outside the bounds of a PR and should be discussed more broadly, please let me know.
No, I think it's good. Thank you.

ianmcook commented:
> The only remaining unaddressed comment from your review about the UUID type in Columnar.rst. However, I actually don't read that one as needing to be replaced _unless_ the intention of _canonical_ extension types is that all Arrow implementations _must_ or _should_ support them. Reading the page on [canonical extensions](https://arrow.apache.org/docs/format/CanonicalExtensions.html) it is unclear to me if arrow implementations _must_ , _should_ , or _may_ support them. If it is _may_, then I think the text of Columnar is actually quite good, though I might modify it as follows:
> 
> > For example, Arrow specifies a canonical extension type that represents a UUID as a `FixedSizeBinary(16)`. Arrow implementations are not required to support canonical extensions, so an implementation that does not support this UUID type will simply interpret it as a `FixedSizeBinary(16)` and pass along the `custom_metadata` in subsequent Arrow protocol messages.

I like your suggested modification. I think it's fine that we refer to a canonical extension type there, but I think it's worth saying explicitly that it's a canonical extension type, which is exactly what your modified text does. 👍

khwilson commented:
Thanks, @ianmcook . Addressed your comments and I think I figured out at least one reason the doctests were failing and have fixed that.

ianmcook commented:
@github-actions crossbow submit preview-docs

ianmcook commented:
@github-actions crossbow submit preview-docs

ianmcook commented:
@khwilson friendly nudge to please sync your fork with upstream main to see if that helps resolve the errors. https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

khwilson commented:
Hi @ianmcook i did do the sync and also setup archery locally to see if I could do some debugging separate from the long GA flow. I’ve resolved one issue (the missing “unregister” call) but another issue popped up: the doctests.ELLIPSIS matcher started failing on *other* tests.

Perhaps I’m missing some local environment variable setup?

ianmcook commented on python/pyarrow/types.pxi:
```diff
@@ -1618,59 +1618,79 @@ cdef class ExtensionType(BaseExtensionType):
     <pyarrow.lib.ExtensionArray object at ...>
-    [
+    -- is_valid: all not null
       ...
```
It looks like one of the test failures is happening because this shows the ellipsis instead of the full output: https://github.com/apache/arrow/actions/runs/10762310741/job/29849821115?pr=43849#step:6:9145

ianmcook commented:
> Hi @ianmcook i did do the sync and also setup archery locally to see if I could do some debugging separate from the long GA flow. I’ve resolved one issue (the missing “unregister” call) but another issue popped up: the doctests.ELLIPSIS matcher started failing on _other_ tests.
> 
> Perhaps I’m missing some local environment variable setup?

Can you try adding doctest directives like `  # doctest: +ELLIPSIS` to fix that issue?
 https://docs.python.org/3/library/doctest.html#directives

khwilson commented:
Hi all. My deepest apologies. I got myself into a bit of trouble syncing with master and accidentally rebased instead of merging. I will close this issue for now to avoid spamming everyone as there is no way I know to unsubscribe you automatically.

ianmcook commented:
No worries, it happens :) Tag me in the new PR and I will review
